File: phases/12-cryptography-and-security/07-macs-and-hmac/code/main.py
"""
MACs and HMAC — Phase 12, Lesson 07
HMAC-SHA256 from scratch, HKDF, length extension attack, timing attack demo.
"""

import hashlib
import hmac as hmac_stdlib
import struct
import time


# ── SHA-256 (minimal, for HMAC internals) ──────────────────────────────────

SHA256_K = [
    0x428a2f98, 0x71374491, 0xb5c0fbcf, 0xe9b5dba5,
    0x3956c25b, 0x59f111f1, 0x923f82a4, 0xab1c5ed5,
    0xd807aa98, 0x12835b01, 0x243185be, 0x550c7dc3,
    0x72be5d74, 0x80deb1fe, 0x9bdc06a7, 0xc19bf174,
    0xe49b69c1, 0xefbe4786, 0x0fc19dc6, 0x240ca1cc,
    0x2de92c6f, 0x4a7484aa, 0x5cb0a9fc, 0x76f988da,
    0x983e5152, 0xa831c66d, 0xb00327c8, 0xbf597fc7,
    0xc6e00bf3, 0xd5a79147, 0x06ca6351, 0x14292967,
    0x27b70a85, 0x2e1b2138, 0x4d2c6dfc, 0x53380d13,
    0x650a7354, 0x766a0abb, 0x81c2c92e, 0x92722c85,
    0xa2bfe8a1, 0xa81a664b, 0xc24b8b70, 0xc76c51a3,
    0xd192e819, 0xd6990624, 0xf40e3585, 0x106aa070,
    0x19a4c116, 0x1e376c08, 0x2748774c, 0x34b0bcb5,
    0x391c0cb3, 0x4ed8aa4a, 0x5b9cca4f, 0x682e6ff3,
    0x748f82ee, 0x78a5636f, 0x84c87814, 0x8cc70208,
    0x90befffa, 0xa4506ceb, 0xbef9a3f7, 0xc67178f2,
]

SHA256_H0 = [
    0x6a09e667, 0xbb67ae85, 0x3c6ef372, 0xa54ff53a,
    0x510e527f, 0x9b05688c, 0x1f83d9ab, 0x5be0cd19,
]

MASK32 = 0xFFFFFFFF


def _rotr32(x, n):
    return ((x >> n) | (x << (32 - n))) & MASK32


def sha256(data: bytes) -> bytes:
    h = list(SHA256_H0)
    msg = bytearray(data)
    msg_len_bits = len(data) * 8
    msg.append(0x80)
    while len(msg) % 64 != 56:
        msg.append(0x00)
    msg += struct.pack(">Q", msg_len_bits)

    for chunk_start in range(0, len(msg), 64):
        chunk = msg[chunk_start:chunk_start + 64]
        w = list(struct.unpack(">16L", chunk))
        for i in range(16, 64):
            s0 = _rotr32(w[i-15], 7) ^ _rotr32(w[i-15], 18) ^ (w[i-15] >> 3)
            s1 = _rotr32(w[i-2], 17) ^ _rotr32(w[i-2], 19) ^ (w[i-2] >> 10)
            w.append((w[i-16] + s0 + w[i-7] + s1) & MASK32)
        a, b, c, d, e, f, g, hh = h
        for i in range(64):
            S1 = _rotr32(e, 6) ^ _rotr32(e, 11) ^ _rotr32(e, 25)
            ch = (e & f) ^ ((~e & MASK32) & g)
            temp1 = (hh + S1 + ch + SHA256_K[i] + w[i]) & MASK32
            S0 = _rotr32(a, 2) ^ _rotr32(a, 13) ^ _rotr32(a, 22)
            maj = (a & b) ^ (a & c) ^ (b & c)
            temp2 = (S0 + maj) & MASK32
            hh = g
            g = f
            f = e
            e = (d + temp1) & MASK32
            d = c
            c = b
            b = a
            a = (temp1 + temp2) & MASK32
        h[0] = (h[0] + a) & MASK32
        h[1] = (h[1] + b) & MASK32
        h[2] = (h[2] + c) & MASK32
        h[3] = (h[3] + d) & MASK32
        h[4] = (h[4] + e) & MASK32
        h[5] = (h[5] + f) & MASK32
        h[6] = (h[6] + g) & MASK32
        h[7] = (h[7] + hh) & MASK32

    return struct.pack(">8L", *h)


# ── HMAC-SHA256 from scratch ─────────────────────────────────────────────

SHA256_BLOCK_SIZE = 64
SHA256_DIGEST_SIZE = 32


def hmac_sha256(key: bytes, message: bytes) -> bytes:
    if len(key) > SHA256_BLOCK_SIZE:
        key = sha256(key)
    key = key + b'\x00' * (SHA256_BLOCK_SIZE - len(key))

    ipad = bytes(k ^ 0x36 for k in key)
    opad = bytes(k ^ 0x5C for k in key)

    inner = sha256(ipad + message)
    return sha256(opad + inner)


# ── HKDF (HMAC-based Extract-and-Expand) ───────────────────────────────

def hkdf_extract(salt: bytes, ikm: bytes) -> bytes:
    return hmac_sha256(salt, ikm)


def hkdf_expand(prk: bytes, info: bytes, length: int) -> bytes:
    n = (length + SHA256_DIGEST_SIZE - 1) // SHA256_DIGEST_SIZE
    okm = b""
    t = b""
    for i in range(1, n + 1):
        t = hmac_sha256(prk, t + info + bytes([i]))
        okm += t
    return okm[:length]


# ── Naive MAC (vulnerable to length extension) ──────────────────────────

def naive_mac(key: bytes, message: bytes) -> bytes:
    return sha256(key + message)


def sha256_length_extension(original_hash: bytes, original_len: int, suffix: bytes) -> bytes:
    """
    Given H(K || M) and len(K || M), compute H(K || M || padding || suffix)
    without knowing K or M.
    """
    original_data_len = original_len
    padded_len = ((original_data_len + 9 + 63) // 64) * 64

    padding = b'\x80'
    padding += b'\x00' * (padded_len - original_data_len - 1 - 8)
    padding += struct.pack(">Q", original_data_len * 8)

    glue_padding = padding[0:(padded_len - original_data_len)]

    total_len = padded_len + len(suffix)
    new_msg = bytearray(len(suffix))
    for i in range(64, len(suffix) + 64):
        pass  # we just need the suffix data

    h = list(struct.unpack(">8L", original_hash))
    extended = suffix
    extended_len = total_len

    new_msg = bytearray(extended)

    msg = bytearray(new_msg)
    msg_len_bits = extended_len + len(suffix)
    msg.append(0x80)
    while len(msg) % 64 != 56:
        msg.append(0x00)
    msg += struct.pack(">Q", extended_len * 8 + len(suffix) * 8)

    result = h
    for chunk_start in range(0, len(msg), 64):
        chunk = msg[chunk_start:chunk_start + 64]
        w = list(struct.unpack(">16L", chunk)) if len(chunk) == 64 else [0]*16
        for i in range(16, 64):
            s0 = _rotr32(w[i-15], 7) ^ _rotr32(w[i-15], 18) ^ (w[i-15] >> 3)
            s1 = _rotr32(w[i-2], 17) ^ _rotr32(w[i-2], 19) ^ (w[i-2] >> 10)
            w.append((w[i-16] + s0 + w[i-7] + s1) & MASK32)
        a, b, c, d, e, f, g, hh = result
        for i in range(64):
            S1 = _rotr32(e, 6) ^ _rotr32(e, 11) ^ _rotr32(e, 25)
            ch = (e & f) ^ ((~e & MASK32) & g)
            temp1 = (hh + S1 + ch + SHA256_K[i] + w[i]) & MASK32
            S0 = _rotr32(a, 2) ^ _rotr32(a, 13) ^ _rotr32(a, 22)
            maj = (a & b) ^ (a & c) ^ (b & c)
            temp2 = (S0 + maj) & MASK32
            hh = g
            g = f
            f = e
            e = (d + temp1) & MASK32
            d = c
            c = b
            b = a
            a = (temp1 + temp2) & MASK32
        result[0] = (result[0] + a) & MASK32
        result[1] = (result[1] + b) & MASK32
        result[2] = (result[2] + c) & MASK32
        result[3] = (result[3] + d) & MASK32
        result[4] = (result[4] + e) & MASK32
        result[5] = (result[5] + f) & MASK32
        result[6] = (result[6] + g) & MASK32
        result[7] = (result[7] + hh) & MASK32

    return struct.pack(">8L", *result)


def sha256_continue_from_state(state_bytes: bytes, remaining_data: bytes, total_processed: int) -> bytes:
    """
    Continue SHA-256 from a known state (the hash digest) with new data,
    given total_processed bytes were already hashed. This is the core
    of the length extension attack.
    """
    h = list(struct.unpack(">8L", state_bytes))
    data = bytearray(remaining_data)
    data.append(0x80)
    total_bits = (total_processed + len(remaining_data)) * 8
    while len(data) % 64 != 56:
        data.append(0x00)
    data += struct.pack(">Q", total_bits)

    for chunk_start in range(0, len(data), 64):
        chunk = data[chunk_start:chunk_start + 64]
        w = list(struct.unpack(">16L", chunk))
        for i in range(16, 64):
            s0 = _rotr32(w[i-15], 7) ^ _rotr32(w[i-15], 18) ^ (w[i-15] >> 3)
            s1 = _rotr32(w[i-2], 17) ^ _rotr32(w[i-2], 19) ^ (w[i-2] >> 10)
            w.append((w[i-16] + s0 + w[i-7] + s1) & MASK32)
        a, b, c, d, e, f, g, hh = h
        for i in range(64):
            S1 = _rotr32(e, 6) ^ _rotr32(e, 11) ^ _rotr32(e, 25)
            ch = (e & f) ^ ((~e & MASK32) & g)
            temp1 = (hh + S1 + ch + SHA256_K[i] + w[i]) & MASK32
            S0 = _rotr32(a, 2) ^ _rotr32(a, 13) ^ _rotr32(a, 22)
            maj = (a & b) ^ (a & c) ^ (b & c)
            temp2 = (S0 + maj) & MASK32
            hh = g
            g = f
            f = e
            e = (d + temp1) & MASK32
            d = c
            c = b
            b = a
            a = (temp1 + temp2) & MASK32
        h[0] = (h[0] + a) & MASK32
        h[1] = (h[1] + b) & MASK32
        h[2] = (h[2] + c) & MASK32
        h[3] = (h[3] + d) & MASK32
        h[4] = (h[4] + e) & MASK32
        h[5] = (h[5] + f) & MASK32
        h[6] = (h[6] + g) & MASK32
        h[7] = (h[7] + hh) & MASK32

    return struct.pack(">8L", *h)


def compute_glue_padding(msg_len: int) -> bytes:
    padded_len = ((msg_len + 1 + 8 + 63) // 64) * 64
    padding = b'\x80'
    padding += b'\x00' * (padded_len - msg_len - 1 - 8)
    padding += struct.pack(">Q", msg_len * 8)
    return padding


def demonstrate_length_extension_attack():
    """
    Demonstrate that H(K || M) is vulnerable to length extension,
    but HMAC(K, M) is NOT.
    """
    key = b'secret_key_1234'
    original_message = b'transfer $100 to Alice'
    extension = b' and $50 to Eve'

    print("=" * 60)
    print("LENGTH EXTENSION ATTACK DEMONSTRATION")
    print("=" * 60)

    naive_tag = naive_mac(key, original_message)
    print(f"\nOriginal message: {original_message}")
    print(f"Naive MAC tag:    {naive_tag.hex()}")

    # Compute glue padding for K || M
    km_len = len(key) + len(original_message)
    glue_padding = compute_glue_padding(km_len)
    total_processed = km_len + len(glue_padding)

    # The attacker sees the tag and knows len(K||M) but NOT K or M
    # They compute H(K || M || glue_padding || extension) from H(K||M)
    forged_tag = sha256_continue_from_state(naive_tag, extension, total_processed)

    # Verify: compute the actual hash of K || M || glue_padding || extension
    full_message = original_message + glue_padding + extension
    actual_hash = sha256(key + full_message)

    print(f"\nForged tag (via length extension):  {forged_tag.hex()}")
    print(f"Actual hash K||M||pad||ext:        {actual_hash.hex()}")
    print(f"Length extension attack works:     {forged_tag == actual_hash}")
    print(f"The attacker forged a valid tag WITHOUT knowing the key!")

    # Now show HMAC resists this
    hmac_tag = hmac_sha256(key, original_message)
    print(f"\nHMAC tag:                          {hmac_tag.hex()}")
    print(f"HMAC resists length extension:     the inner hash hides state")

    # Try the same trick on HMAC — it fails
    hmac_forged = sha256_continue_from_state(hmac_tag, extension, SHA256_BLOCK_SIZE + SHA256_DIGEST_SIZE)
    hmac_actual = sha256(bytes(k ^ 0x5C for k in (key + b'\x00' * (SHA256_BLOCK_SIZE - len(key)))) +
                         sha256(bytes(k ^ 0x36 for k in (key + b'\x00' * (SHA256_BLOCK_SIZE - len(key)))) +
                                original_message + extension))
    print(f"Forcing length extension on HMAC forged: {hmac_forged.hex()}")
    print(f"Actual HMAC of extended message:          {hmac_actual.hex()}")
    print(f"Length extension on HMAC works:           {hmac_forged == hmac_actual}")
    print("HMAC is NOT vulnerable to length extension attacks!")


# ── HKDF Test Vectors (RFC 5869) ───────────────────────────────────────

def test_hkdf():
    print("\n" + "=" * 60)
    print("HKDF TEST VECTORS (RFC 5869, Test Case 1)")
    print("=" * 60)

    ikm = bytes.fromhex("0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b")
    salt = bytes.fromhex("000102030405060708090a0b0c")
    info = bytes.fromhex("f0f1f2f3f4f5f6f7f8f9")
    l = 42

    prk = hkdf_extract(salt, ikm)
    okm = hkdf_expand(prk, info, l)

    expected_prk = bytes.fromhex(
        "077709362c2e32df0ddc3f0dc47bba63"
        "90b6c73bb50f9c3122ca8e4f3df5970"
    )
    expected_okm = bytes.fromhex(
        "3cb25f25faacd57a90434f64d0362f2a"
        "2d2d0a90cf1a5a4c5db5418eac4aee2"
        "3d1a29b1a4d2f439de3ffbe3f4af705"
        "cf5d7bba"
    )

    print(f"PRK computed:  {prk.hex()}")
    print(f"PRK expected:  {expected_prk.hex()}")
    print(f"PRK matches:   {prk == expected_prk}")
    print(f"\nOKM computed:  {okm.hex()}")
    print(f"OKM expected:  {expected_okm.hex()}")
    print(f"OKM matches:   {okm == expected_okm}")


# ── HMAC-SHA256 Test Vectors (RFC 4231) ──────────────────────────────────

def test_hmac_vectors():
    print("\n" + "=" * 60)
    print("HMAC-SHA256 TEST VECTORS (RFC 4231)")
    print("=" * 60)

    test_cases = [
        {
            "key": bytes.fromhex("0b" * 20),
            "data": b"Hi There",
            "expected": "b0344c703b8ec6cf82e8b4d0394b4b0a3b6f85d7b0f0303f7afab2403f063029",
        },
        {
            "key": b"Jefe",
            "data": b"what do ya want for nothing?",
            "expected": "5bdcc146bf60754e6a042426089575c75a003f089d2739839dec58b964ec3843",
        },
        {
            "key": bytes.fromhex("aa" * 20),
            "data": bytes.fromhex("dd" * 50),
            "expected": "773ea91e36800e46854db8ebd09181a72986033167f883d0289f18758fdb3822",
        },
    ]

    all_pass = True
    for i, tc in enumerate(test_cases, 1):
        result = hmac_sha256(tc["key"], tc["data"])
        expected = bytes.fromhex(tc["expected"])
        match = result == expected
        all_pass = all_pass and match
        print(f"\n  Test {i}: key={tc['key'][:8].hex()}..., "
              f"data={tc['data'][:16]}..., match={match}")
        if not match:
            print(f"    Got:      {result.hex()}")
            print(f"    Expected: {tc['expected']}")

    # Cross-verify with stdlib
    key = b"test_key_12345678"
    msg = b"Hello, HMAC world!"
    ours = hmac_sha256(key, msg)
    theirs = hmac_stdlib.new(key, msg, hashlib.sha256).digest()
    print(f"\n  Cross-verify with hashlib: {ours == theirs}")
    all_pass = all_pass and (ours == theirs)

    print(f"\n  All HMAC test vectors pass: {all_pass}")


# ── Timing Attack on MAC Comparison ────────────────────────────────────────

def insecure_compare(a: bytes, b: bytes) -> bool:
    """VULNERABLE: returns early on first mismatch, leaking timing."""
    if len(a) != len(b):
        return False
    for x, y in zip(a, b):
        if x != y:
            return False
    return True


def constant_time_compare(a: bytes, b: bytes) -> bool:
    """SAFE: always scans all bytes regardless of match position."""
    if len(a) != len(b):
        return False
    result = 0
    for x, y in zip(a, b):
        result |= x ^ y
    return result == 0


def demonstrate_timing_attack():
    """
    Show that byte-by-byte comparison leaks information through timing.
    An attacker can recover the tag byte by byte by measuring response times.
    """
    print("\n" + "=" * 60)
    print("TIMING ATTACK ON MAC COMPARISON")
    print("=" * 60)

    key = b"my_secret_key_for_demo_12345"
    message = b"Important financial transaction"
    correct_tag = hmac_sha256(key, message)

    print(f"\nCorrect tag: {correct_tag.hex()[:32]}...")

    # Measure timing for tags with increasing prefix matches
    print("\n  Insecure comparison — timing leaks prefix match length:")
    print(f"  {'Prefix match':<16} {'Avg time (µs)':<16} {'Insecure result'}")

    for prefix_len in range(0, 33, 4):
        forged = bytearray(correct_tag[:prefix_len]) + bytearray([0xFF] * (32 - prefix_len))
        forged = bytes(forged)

        times = []
        for _ in range(10000):
            start = time.perf_counter_ns()
            insecure_compare(correct_tag, forged)
            elapsed = time.perf_counter_ns() - start
            times.append(elapsed)

        avg_ns = sum(times) / len(times)
        result = insecure_compare(correct_tag, forged)
        print(f"  {prefix_len:>3} bytes       {avg_ns:>10.0f}       {result}")

    print("\n  Note: timing increases with prefix match length — attacker can")
    print("  recover the tag byte by byte by observing which猜测 maximizes delay.")

    print("\n  Constant-time comparison — no timing leak:")
    for prefix_len in [0, 16, 32]:
        forged = bytearray(correct_tag[:prefix_len]) + bytearray([0xFF] * (32 - prefix_len))
        forged = bytes(forged)

        times = []
        for _ in range(10000):
            start = time.perf_counter_ns()
            constant_time_compare(correct_tag, forged)
            elapsed = time.perf_counter_ns() - start
            times.append(elapsed)

        avg_ns = sum(times) / len(times)
        result = constant_time_compare(correct_tag, forged)
        print(f"  {prefix_len:>3} bytes match   {avg_ns:>10.0f}       {result}")

    print("\n  Python hmac.compare_digest() provides constant-time comparison.")
    print("  Rust: use subtle::ConstantTimeEq.")


# ── Message Integrity Demo ────────────────────────────────────────────────

def demonstrate_message_integrity():
    print("\n" + "=" * 60)
    print("MESSAGE INTEGRITY AND AUTHENTICATION DEMO")
    print("=" * 60)

    key = b"shared_secret_between_alice_bob"
    original_message = b"Transfer $500 to account #1234"
    tampered_message = b"Transfer $500 to account #5678"

    original_tag = hmac_sha256(key, original_message)
    tampered_tag = hmac_sha256(key, tampered_message)

    print(f"\nOriginal message: {original_message}")
    print(f"Original tag:      {original_tag.hex()}")
    print(f"\nTampered message:  {tampered_message}")
    print(f"Tampered tag:      {tampered_tag.hex()}")

    print(f"\nVerify original (tag matches): {hmac_sha256(key, original_message) == original_tag}")
    print(f"Verify tampered (tag differs): {hmac_sha256(key, tampered_message) == original_tag}")

    print(f"\nAttack scenario: Eve intercepts and modifies the message.")
    print(f"She cannot produce a valid tag for her modified message without the key.")
    print(f"Bob detects the forgery: tag mismatch.")


# ── Main ──────────────────────────────────────────────────────────────────

def main():
    print("MACs and HMAC — Phase 12, Lesson 07")
    print("=" * 60)

    test_hmac_vectors()
    test_hkdf()
    demonstrate_length_extension_attack()
    demonstrate_message_integrity()
    demonstrate_timing_attack()

    print("\n" + "=" * 60)
    print("All demonstrations complete.")
    print("=" * 60)


if __name__ == "__main__":
    main()

File: phases/12-cryptography-and-security/07-macs-and-hmac/code/main.rs
//! MACs and HMAC — Phase 12, Lesson 07
//! HMAC-SHA256 and CBC-MAC implementations with constant-time verification.

use sha2::{Digest, Sha256};

const SHA256_BLOCK_SIZE: usize = 64;
const SHA256_DIGEST_SIZE: usize = 32;

// ── HMAC-SHA256 ──────────────────────────────────────────────────────────

fn hmac_sha256(key: &[u8], message: &[u8]) -> [u8; 32] {
    let padded_key = if key.len() > SHA256_BLOCK_SIZE {
        let h = Sha256::digest(key);
        let mut padded = vec![0u8; SHA256_BLOCK_SIZE];
        padded[..32].copy_from_slice(&h);
        padded
    } else {
        let mut padded = vec![0u8; SHA256_BLOCK_SIZE];
        padded[..key.len()].copy_from_slice(key);
        padded
    };

    let ipad: Vec<u8> = padded_key.iter().map(|k| k ^ 0x36).collect();
    let opad: Vec<u8> = padded_key.iter().map(|k| k ^ 0x5C).collect();

    let inner_hash = {
        let mut hasher = Sha256::new();
        hasher.update(&ipad);
        hasher.update(message);
        hasher.finalize().to_vec()
    };

    let mut outer_hasher = Sha256::new();
    outer_hasher.update(&opad);
    outer_hasher.update(&inner_hash);

    let result = outer_hasher.finalize();
    let mut tag = [0u8; 32];
    tag.copy_from_slice(&result);
    tag
}

// ── CBC-MAC (AES-128) ──────────────────────────────────────────────────

fn aes128_encrypt(block: &[u8; 16], key: &[u8; 16]) -> [u8; 16] {
    use aes::cipher::{BlockEncrypt, KeyInit};
    let cipher = aes::Aes128::new(key.into());
    let mut b = aes::cipher::generic_array::GenericArray::clone_from_slice(block);
    cipher.encrypt_block(&mut b);
    let mut out = [0u8; 16];
    out.copy_from_slice(&b);
    out
}

fn xor_blocks(a: &[u8; 16], b: &[u8; 16]) -> [u8; 16] {
    let mut out = [0u8; 16];
    for i in 0..16 {
        out[i] = a[i] ^ b[i];
    }
    out
}

fn cbc_mac(key: &[u8; 16], message: &[u8]) -> [u8; 16] {
    assert!(
        message.len() % 16 == 0,
        "CBC-MAC requires message length to be a multiple of 16 bytes"
    );

    let mut chain = [0u8; 16];
    for chunk in message.chunks(16) {
        let mut block = [0u8; 16];
        block.copy_from_slice(chunk);
        chain = aes128_encrypt(&xor_blocks(&chain, &block), key);
    }
    chain
}

// ── Constant-time comparison ────────────────────────────────────────────

fn constant_time_eq(a: &[u8], b: &[u8]) -> bool {
    if a.len() != b.len() {
        return false;
    }
    let mut result: u8 = 0;
    for (x, y) in a.iter().zip(b.iter()) {
        result |= x ^ y;
    }
    result == 0
}

// ── Test vectors ────────────────────────────────────────────────────────

fn test_hmac_sha256_vectors() {
    println!("HMAC-SHA256 Test Vectors (RFC 4231)");
    println!("{}", "-".".repeat(40));

    let test_cases = [
        (
            vec![0x0b; 20],
            b"Hi There".to_vec(),
            "b0344c703b8ec6cf82e8b4d0394b4b0a3b6f85d7b0f0303f7afab2403f063029",
        ),
        (
            b"Jefe".to_vec(),
            b"what do ya want for nothing?".to_vec(),
            "5bdcc146bf60754e6a042426089575c75a003f089d2739839dec58b964ec3843",
        ),
        (
            vec![0xaa; 20],
            vec![0xdd; 50],
            "773ea91e36800e46854db8ebd09181a72986033167f883d0289f18758fdb3822",
        ),
        (
            vec![0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f, 0x10, 0x11, 0x12, 0x13, 0x14, 0x15, 0x16, 0x17, 0x18, 0x19],
            vec![0xcd; 50],
            "d730594dd1672e82f7c0208f9762788201c12b5bdac24bc25f6f5933a1aaffed",
        ),
    ];

    let mut all_pass = true;
    for (i, (key, data, expected_hex)) in test_cases.iter().enumerate() {
        let tag = hmac_sha256(key, data);
        let tag_hex: String = tag.iter().map(|b| format!("{:02x}", b)).collect();
        let pass = tag_hex == *expected_hex;
        all_pass = all_pass && pass;
        println!("  Test {}: key_len={:>3} data_len={:>3} pass={}", i + 1, key.len(), data.len(), pass);
        if !pass {
            println!("    Got:      {}", tag_hex);
            println!("    Expected: {}", expected_hex);
        }
    }
    println!("  All HMAC-SHA256 vectors pass: {}\n", all_pass);
}

fn test_cbc_mac() {
    println!("CBC-MAC Test Vectors");
    println!("{}", "-".".repeat(40));

    // NIST SP 800-38B CMAC test vectors (also valid for CBC-MAC on single-block messages)
    let key: [u8; 16] = [
        0x2b, 0x7e, 0x15, 0x16, 0x28, 0xae, 0xd2, 0xa6,
        0xab, 0xf7, 0x15, 0x88, 0x09, 0xcf, 0x4f, 0x3c,
    ];
    let msg: [u8; 16] = [
        0x6b, 0xc1, 0xbe, 0xe2, 0x2e, 0x40, 0x9f, 0x96,
        0xe9, 0x3d, 0x7e, 0x11, 0x73, 0x93, 0x17, 0x2a,
    ];
    let tag = cbc_mac(&key, &msg);
    let tag_hex: String = tag.iter().map(|b| format!("{:02x}", b)).collect();
    println!("  Key:   2b7e151628aed2a6abf7158809cf4f3c");
    println!("  Msg:   6bc1bee22e409f96e93d7e117393172a");
    println!("  Tag:   {}", tag_hex);
    println!("  (CBC-MAC on single-block message)");

    // Verify with self-consistency
    let tag2 = cbc_mac(&key, &msg);
    println!("  Self-consistent: {}\n", tag == tag2);
}

fn test_constant_time_comparison() {
    println!("Constant-Time Comparison");
    println!("{}", "-".repeat(40));

    let key = b"my_secret_key_for_demo_12345";
    let message = b"Important financial transaction";
    let tag = hmac_sha256(key, message);

    // Verify correct tag
    let recomputed = hmac_sha256(key, message);
    println!("  Correct tag verification: {}", constant_time_eq(&tag, &recomputed));

    // Detect tampered message
    let tampered = b"Important financial transactiom";
    let tampered_tag = hmac_sha256(key, tampered);
    println!("  Tampered message detected:  {}", !constant_time_eq(&tag, &tampered_tag));

    // Single-bit difference
    let mut forged = tag;
    forged[0] ^= 0x01;
    println!("  Single-bit flip detected:    {}\n", !constant_time_eq(&tag, &forged));
}

fn demonstrate_message_integrity() {
    println!("Message Integrity and Authentication");
    println!("{}", "-".repeat(40));

    let key = b"shared_secret_between_alice_bob";
    let original = b"Transfer $500 to account #1234";
    let tampered = b"Transfer $500 to account #5678";

    let original_tag = hmac_sha256(key, original);

    println!("  Original: {}", String::from_utf8_lossy(original));
    println!("  Tag:      {}", hex::encode(original_tag));

    println!("\n  Tampered: {}", String::from_utf8_lossy(tampered));
    println!("  Tag:      {}", hex::encode(hmac_sha256(key, tampered)));

    println!("\n  Verify original: tag match = {}", constant_time_eq(&original_tag, &hmac_sha256(key, original)));
    println!("  Verify tampered: tag match = {}", constant_time_eq(&original_tag, &hmac_sha256(key, tampered)));
    println!("  Eve cannot forge a valid tag without the key.\n");
}

fn print_comparison_table() {
    println!("MAC Comparison Table");
    println!("{}", "-".repeat(40));
    println!("  {:<12} {:<10} {:<12} {:<16} {:<10}", "Scheme", "Basis", "Key Reuse", "Var Length", "Standard");
    println!("  {:<12} {:<10} {:<12} {:<16} {:<10}", "-----", "-----", "---------", "----------", "--------");
    println!("  {:<12} {:<10} {:<12} {:<16} {:<10}", "HMAC", "Hash", "Yes", "Yes", "RFC 2104");
    println!("  {:<12} {:<10} {:<12} {:<16} {:<10}", "CBC-MAC", "Cipher", "Yes", "No (fixed)", "SP 800-38A");
    println!("  {:<12} {:<10} {:<12} {:<16} {:<10}", "CMAC", "Cipher", "Yes", "Yes", "SP 800-38B");
    println!("  {:<12} {:<10} {:<12} {:<16} {:<10}", "Poly1305", "Polynomial", "No (1-time)", "Yes", "RFC 8439");
    println!();
}

fn main() {
    println!("MACs and HMAC \u{2014} Phase 12, Lesson 07");
    println!("{}\n", "=".repeat(50));

    test_hmac_sha256_vectors();
    test_cbc_mac();
    test_constant_time_comparison();
    demonstrate_message_integrity();
    print_comparison_table();

    println!("All tests complete.");
}